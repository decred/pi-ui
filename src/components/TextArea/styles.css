.textArea {
  font-family: var(--font-family-text);
  font-weight: var(--font-weight-regular);
  max-width: 100%;
  min-width: 100%;
  padding: 2rem;
  min-height: 20rem;
  border: 1px solid var(--text-input-color);
  line-height: 1.5;
  font-size: var(--font-size-regular);
  background: var(--card-background);
  -webkit-appearance: none;
  color: var(--text-color);
}

.textAreaWrapper {
  width: 100%;
  position: relative;
}

.textAreaWrapper:not(:first-child) {
  margin-top: 2.2rem;
}

.textArea:focus,
.textArea:not(:placeholder-shown) {
  top: -2rem;
  font-size: var(--font-size-small);
  border-bottom: 0.1rem solid var(--color-primary);
}

.errorMsg {
  visibility: hidden;
  color: var(--color-orange);
  font-size: var(--font-size-small);
  min-height: var(--spacing-1);
  line-height: var(--spacing-1);
  margin-top: 1px;
}

.errorMsgActive {
  visibility: visible;
}

.errorIcon {
  display: none;
  color: var(--color-orange);
  position: absolute;
  right: 0.5rem;
  top: 0.5rem;
  font-size: var(--font-size-large);
}

.errorIconActive {
  display: block;
}

/* override chrome's autocomplete styling */

@keyframes autofill {
  to {
    color: white;
    background: transparent;
  }
}

.textArea:-webkit-autofill {
  transition-delay: 99999s;
  -webkit-transition-delay: 99999s;
  animation-name: "autofill";
  -webkit-animation-name: "autofill";
  animation-fill-mode: both;
  -webkit-animation-fill-mode: both;
  -webkit-text-fill-color: var(--text-color) !important;
}
