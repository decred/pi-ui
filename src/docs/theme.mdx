---
name: Theming
route: /theming
---

# Theming

## Create a custom theme

The recommended way to create a custom theme is to create a file named `theme.js`, you can add your custom values and override existing ones.

```js
// theme.js

export default {
  "color-primary": "red",
  "btn-color": "var(--color-primary)"
};
```

## Use your theme in your application

Pi-ui exposes a `ThemeProvider` that accepts a `themes` object, `defaultThemeName` as string and `fonts` array.

```js
import React from "react";
import { render } from "react-dom";
import { ThemeProvider, defaultLightTheme } from "pi-ui";

import customTheme from "./theme";
import SourceSansProLight from "../assets/fonts/source-sans-pro/SourceSansPro-Light.ttf";
import SourceSansProRegular from "../assets/fonts/source-sans-pro/SourceSansPro-Regular.ttf";
import SourceSansProSemiBold from "../assets/fonts/source-sans-pro/SourceSansPro-SemiBold.ttf";

const fonts = [
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProLight}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-light"], // 300
    "font-style": "normal",
    "font-display": "swap"
  },
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProRegular}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-regular"], // 400
    "font-style": "normal",
    "font-display": "swap"
  },
  {
    "font-family": "Source Sans Pro",
    src: `url(${SourceSansProSemiBold}) format("truetype")`,
    "font-weight": defaultLightTheme["font-weight-semi-bold"], // 600
    "font-style": "normal",
    "font-display": "swap"
  }
];

const App = () => {
  return (
    <ThemeProvider
      themes={{ light: defaultLightTheme, customTheme }}
      defaultThemeName="customTheme"
      fonts={fonts}>
      {children}
    </ThemeProvider>
  );
};

render(<App />, document.getElementById("root"));
```

## Default theme

There are two default themes, `defaultLightTheme` and `defaultDarkTheme`. They are both name exported from pi-ui.

```js
import { defaultLightTheme, defaultDarkTheme } from "pi-ui";
```

## UseTheme hook in deep children

`useTheme` exposes three properties `theme`, `themeName`, `setThemeName` which can be used in deep children.

```js
import React from "react";
import { useTheme, Button, getThemeProperty } from "pi-ui";

const DeepChild = () => {
  const { themeName, setThemeName, theme, DEFAULT_LIGHT_THEME_NAME,
    DEFAULT_DARK_THEME_NAME } = useTheme();

  const colorGreen = getThemeProperty(theme, "color-green"); // get theme property 
  
  const handleToggleTheme = () => {
    if (themeName === DEFAULT_LIGHT_THEME_NAME) {
      setThemeName(DEFAULT_LIGHT_THEME_NAME);
    } else {
      setThemeName(DEFAULT_LIGHT_THEME_NAME);
    }
  }

  return (
    <Button onClick={handleToggleTheme}>Toggle theme</Button>
  )
};
```
