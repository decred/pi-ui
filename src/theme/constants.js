/*
 * The following constants are used throughout the application to determine
 * if the current active theme is default light or dark theme
 */
export const DEFAULT_LIGHT_THEME_NAME = "light";
export const DEFAULT_DARK_THEME_NAME = "dark";
