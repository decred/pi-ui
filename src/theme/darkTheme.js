import lightTheme from "./lightTheme";

const darkTheme = {
  ...lightTheme,

  "color-primary-dark": "#7DA7D9",
  "color-primary-darker": "#1F325F",
  "color-primary-darkest": "#1B2B54",
  "color-dark": "#091440",
  "color-gray-blue": "#6F98D3",

  "color-powder-blue": "#B7DEEE",
  "color-astronaut-blue": "#29417A",

  "html-bg": "var(--color-primary-darkest)",

  "text-input-color": "var(--color-gray-blue)",
  "container-color": "var(--color-primary-darkest)",
  "header-border-color": "#1d1d2f",
  "header-color": "var(--color-primary-darker)",
  "topbanner-color": "var(--card-background)",
  "main-color": "#27283d",
  "input-border-color": "var(--color-gray-blue)",
  "active-nav-border-color": "var(--color-primary-light)",
  "separator-color": "var(--color-primary-darkest)",
  "subtitle-color": "#3C62B0",
  "card-background": "#223767",
  "dimmed-card-background": "#1F2F5A",
  "success-icon-background-color": "#527CC9",
  "success-icon-checkmark-color": "var(--color-primary-darkest)",

  "btn-disabled-background-color": "var(--color-primary-darker)",
  "btn-disabled-text-color": "var(--info-message-background)",

  "comment-like-color": "var(--color-gray-blue)",
  "comment-like-color-active": "var(--color-powder-blue)",
  "comment-censored-background": "var(--dimmed-card-background)",
  "comment-new-background": "var(--color-primary-darker)",

  "text-color": "#99C1E3",
  "text-color-light": "var(--color-white)",
  "text-heading-color": "var(--color-powder-blue)",
  "text-heading2-color": "var(--text-heading-color)",
  "text-heading3-color": "#E9F8FE",
  "text-secondary-color": "var(--color-primary-dark)",

  "tab-text-color": "var(--color-gray-blue)",
  "tab-text-active-color": "#E9F8FE",
  "tab-count-background": "#2F4D8C",

  "icon-color": "#36579E",
  "icon-background-color": "var(--text-color)",
  "icon-hover-color": "var(--color-primary-dark)",

  "table-header-background": "var(--dimmed-card-background)",
  "table-cell-border-color": "var(--separator-color)",
  "table-page-arrow-background": "var(--card-background)",
  "table-page-arrow-text-color": "var(--tab-text-active-color)",
  "table-active-page-arrow-text-color": "var(--text-color)",
  "table-active-page-arrow-background": "var(--tab-count-background)",

  "md-table-cell-border-color": "var(--color-primary-dark)",
  "md-table-odd-row-background": "var(--color-primary-darker)",
  "md-table-even-row-background": "var(--color-primary-darkest)",

  "toggle-bar-color": "var(--color-primary-darkest)",

  "dropdown-item-hover-color": "var(--color-powder-blue)",
  "dropdown-arrow-color": "var(--text-color)",

  "select-anchor-color": "var(--text-color)",

  "copyable-text-background-color": "var(--color-primary-darkest)",

  "modal-close-color": "var(--text-color)",

  "tooltip-background-color": "var(--color-astronaut-blue)",

  "checkbox-stroke-color": "var(--color-gray-blue)",
  "checkbox-hovered-background-color": "var(--color-primary-darkest)",

  "info-message-background": "#152042",
  "blocked-message-background": "var(--info-message-background)",
  "error-message-background": "var(--info-message-background)",
  "error-message-color": "var(--color-astronaut-blue)",
  "blocked-message-color": "var(--error-message-color)",
  "warning-message-background": "var(--info-message-background)",
  "warning-message-color": "var(--color-astronaut-blue)",

  "datasheet-read-only-text-color": "var(--color-gray)",
  "datasheet-read-only-background": "var(--color-primary-darkest)",
  "datasheet-border-color": "var(--color-gray-blue)"
};

export default darkTheme;
