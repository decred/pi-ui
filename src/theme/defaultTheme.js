import lightTheme from "./lightTheme";

function applyDefaultTheme() {
  Object.keys(lightTheme).forEach((key) => {
    document.documentElement.style.setProperty(`--${key}`, lightTheme[key]);
  });
}

applyDefaultTheme();
