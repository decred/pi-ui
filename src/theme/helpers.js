/**
 * Returns the final value of a given theme key
 * @param {Object} theme
 * @param {String} key
 */
export const getThemeProperty = (theme, key) => {
  if (!theme[key]) {
    return new Error("Please, provide a valid color key!");
  }
  const value = theme[key];
  const regex = /^var\(--(.*)\)/gm;
  const match = regex.exec(value);

  const isVariable = match && match.length > 0;
  if (isVariable) {
    return getThemeProperty(theme, match[1]);
  }

  return value;
};
