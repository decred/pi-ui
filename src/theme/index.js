export { default as defaultDarkTheme } from "./darkTheme";
export { default as defaultLightTheme } from "./lightTheme";
export { default as useFont } from "./useFont";
export * from "./ThemeProvider";
export * from "./helpers";
