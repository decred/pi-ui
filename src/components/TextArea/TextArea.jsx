import PropTypes from "prop-types";
import React from "react";
import { classNames } from "../../utils";
import Icon from "../Icon/Icon.jsx";
import styles from "./styles.css";

const TextArea = ({
  type,
  id,
  error,
  wrapperClassNames,
  inputClassNames,
  wrapped,
  ...props
}) => {
  return !wrapped ? (
    <div
      className={classNames(
        styles.textAreaWrapper,
        wrapperClassNames,
        wrapped && styles.fullHeight
      )}>
      <textarea
        id={id}
        placeholder=""
        className={classNames(
          styles.textArea,
          error && styles.textAreaError,
          wrapped && styles.fullHeight,
          inputClassNames
        )}
        type={type}
        {...props}
      />
      <Icon
        type="alert"
        backgroundColor="#ed6d47"
        iconColor="#feb8a5"
        className={classNames(
          styles.errorIcon,
          error && styles.errorIconActive
        )}
      />
      <p
        className={classNames(
          styles.errorMsg,
          error && styles.errorMsgActive,
          wrapped && styles.noErrorMsg
        )}>
        {error}
      </p>
    </div>
  ) : (
    <textarea
      id={id}
      placeholder=""
      className={classNames(
        styles.textArea,
        error && styles.textAreaError,
        inputClassNames
      )}
      type={type}
      {...props}
    />
  );
};

TextArea.propTypes = {
  type: PropTypes.string,
  id: PropTypes.string.isRequired,
  error: PropTypes.string,
  wrapperClassNames: PropTypes.string,
  inputClassNames: PropTypes.string,
  placeholder: PropTypes.string,
  wrapped: PropTypes.bool
};

TextArea.defaultProps = {
  type: "text",
  placeholder: ""
};

export default TextArea;
