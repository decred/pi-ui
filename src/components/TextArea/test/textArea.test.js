import React from "react";
import TextArea from "../TextArea";
import { create } from "react-test-renderer";
import { defaultLightTheme, ThemeProvider } from "../../../theme";

describe("TextArea component", () => {
  test("Matches the snapshot", () => {
    const textArea = create(
      <ThemeProvider
        themes={{ light: defaultLightTheme }}
        defaultThemeName="light">
        <TextArea id="test" />
      </ThemeProvider>
    );
    expect(textArea.toJSON()).toMatchSnapshot();
  });
});
