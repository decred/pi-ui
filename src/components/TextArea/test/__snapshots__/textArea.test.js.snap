// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextArea component Matches the snapshot 1`] = `
<div
  className="textAreaWrapper"
>
  <textarea
    className="textArea"
    id="test"
    placeholder=""
    type="text"
  />
  <svg
    className="errorIcon"
    height="16"
    viewBox="0 0 450 450"
    width="16"
  >
    <path
      d="M225.24,24.86c-110.46,0-200,89.55-200,200s89.54,200,200,200,200-89.55,200-200S335.69,24.86,225.24,24.86Z"
      fill="#ed6d47"
    />
    <rect
      fill="none"
    />
    <path
      d="M188.2,313c0-21.92,15.7-36.84,36.8-36.84s36.8,14.92,36.8,36.84-15.7,37.47-36.8,37.47S188.2,335,188.2,313Zm11.56-164.61L198,99.49H252l-1.8,48.94-7.5,103.67H207.26Z"
      fill="#feb8a5"
    />
  </svg>
  <p
    className="errorMsg"
  />
</div>
`;
